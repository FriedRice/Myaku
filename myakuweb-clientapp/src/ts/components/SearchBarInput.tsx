/** @module Search bar input component  */

import { MEDIUM_VIEWPORT_MIN_WIDTH } from '../viewport';
import React from 'react';

interface SearchBarInputProps {
    searchQuery: string;
    onChange: (searchQuery: string) => void;
}
type Props = SearchBarInputProps;

interface SearchBarInputState {
    placeholder: string;
}
type State = SearchBarInputState;

// Search box placeholder text adjusts based on viewport size
const FULL_SEARCH_PLACEHOLDER = 'Japanese word, set phrase, idiom, etc.';
const SHORT_SEARCH_PLACEHOLDER = 'Japanese word, phrase, etc.';

class SearchBarInput extends React.Component<Props, State> {
    constructor(props: Props) {
        super(props);
        this.bindEventHandlers();
        this.state = {
            placeholder: this.getViewportPlaceholder(),
        };
    }

    componentDidMount(): void {
        window.addEventListener('resize', this.updatePlaceholder);
    }

    componentWillUnmount(): void {
        window.removeEventListener('resize', this.updatePlaceholder);
    }

    bindEventHandlers(): void {
        this.updatePlaceholder = this.updatePlaceholder.bind(this);
        this.handleInputClear = this.handleInputClear.bind(this);
        this.handleInputChange = this.handleInputChange.bind(this);
    }

    getViewportPlaceholder(): string {
        if (window.innerWidth >= MEDIUM_VIEWPORT_MIN_WIDTH) {
            return FULL_SEARCH_PLACEHOLDER;
        } else {
            return SHORT_SEARCH_PLACEHOLDER;
        }
    }

    updatePlaceholder(): void {
        this.setState({
            placeholder: this.getViewportPlaceholder(),
        });
    }

    handleInputClear(): void {
        this.props.onChange('');
    }

    handleInputChange(event: React.FormEvent<HTMLInputElement>): void {
        this.props.onChange(event.currentTarget.value);
    }

    render(): React.ReactElement {
        return (
            <div className='search-bar'>
                <input
                    className='search-input'
                    id='search-input'
                    type='text'
                    name='q'
                    aria-label='Search input'
                    value={this.props.searchQuery}
                    placeholder={this.state.placeholder}
                    onChange={this.handleInputChange}
                />
                <button
                    className='search-clear'
                    type='button'
                    aria-label='Search clear'
                    onClick={this.handleInputClear}
                >
                    <i className='fa fa-times'></i>
                </button>
                <button
                    className='search-submit'
                    type='submit'
                    aria-label='Search submit button'
                >
                    <i className='fa fa-search'></i>
                </button>
            </div>
        );
    }
}

export default SearchBarInput;
