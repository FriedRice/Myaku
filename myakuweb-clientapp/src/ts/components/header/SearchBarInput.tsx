/**
 * Search bar input component with submit and input clear buttons.
 * @module ts/components/header/SearchBarInput
 */

import React from 'react';
import { ViewportSize } from 'ts/app/viewport';
import { useCallback } from 'react';
import useInputChangeHandler from 'ts/hooks/useInputChangeHandler';
import useViewportReactiveValue from 'ts/hooks/useViewportReactiveValue';

interface SearchBarInputProps {
    searchQuery: string;
    loading: boolean;
    maxQueryLength: number;
    errorValueSubmitted: boolean;
    onChange: (searchQuery: string) => void;
}
type Props = SearchBarInputProps;

const enum ErrorState {
    OK,
    Warning,
    Error
}

// Search box placeholder text adjusts based on viewport size
const FULL_SEARCH_PLACEHOLDER = 'Japanese word, set phrase, idiom, etc.';
const SHORT_SEARCH_PLACEHOLDER = 'Japanese word, phrase, etc.';
const DEFAULT_PLACEHOLDER = SHORT_SEARCH_PLACEHOLDER;
const VIEWPORT_PLACEHOLDERS = {
    [ViewportSize.Small]: FULL_SEARCH_PLACEHOLDER,
};


function useSearchInput(
    currentValue: string, handleChange: (value: string) => void,
    loading: boolean, errorState: ErrorState
): React.ReactElement {
    const placeholder = useViewportReactiveValue(
        DEFAULT_PLACEHOLDER, VIEWPORT_PLACEHOLDERS
    );

    var classList = ['search-input'];
    if (errorState === ErrorState.Warning) {
        classList.push('warning-border');
    } else if (errorState === ErrorState.Error) {
        classList.push('error-border');
    }

    return (
        <input
            className={classList.join(' ')}
            id='search-input'
            type='text'
            name='q'
            aria-label='Search input'
            value={currentValue}
            placeholder={placeholder}
            readOnly={loading}
            onChange={useInputChangeHandler(handleChange)}
        />
    );
}

function useSearchClearButton(
    handleInputChange: (value: string) => void, loading: boolean
): React.ReactElement {
    const handleInputClear = useCallback(
        function(event: React.SyntheticEvent<HTMLButtonElement>): void {
            event.currentTarget.blur();
            handleInputChange('');
        },
        [handleInputChange]
    );

    return (
        <button
            className='search-clear'
            type='button'
            aria-label='Search clear'
            disabled={loading}
            onClick={handleInputClear}
        >
            <i className='fa fa-times'></i>
        </button>
    );
}

function getSearchSubmitButton(loading: boolean): React.ReactElement {
    var icon = <i className='fa fa-search'></i>;
    if (loading) {
        icon = <div className='content-loading-query-spinner'></div>;
    }
    return (
        <button
            className='search-submit'
            type='submit'
            aria-label='Search submit button'
            disabled={loading}
        >
            {icon}
        </button>
    );
}

function getNoQueryErrorElement(className: string): React.ReactElement {
    return <p className={className}>No search query entered</p>;
}

function getQueryTooLongErrorElement(
    className: string, inputtedQuery: string, maxQueryLength: number,
    errorState: ErrorState
): React.ReactElement {
    var action = 'Inputted';
    if (errorState === ErrorState.Error) {
        action = 'Submitted';
    }
    return (
        <p className={className}>
            <span>{`${action} search query is too long `}</span>
            <span>
                {`(${inputtedQuery.length} / ${maxQueryLength} characters)`}
            </span>
        </p>
    );
}

function getErrorMessageElement(
    inputtedQuery: string, maxQueryLength: number, errorState: ErrorState
): React.ReactElement | null {
    var className;
    if (errorState == ErrorState.Error) {
        className = 'input-error-text';
    } else {
        className = 'input-warning-text';
    }

    if (inputtedQuery.length === 0) {
        return getNoQueryErrorElement(className);
    } else if (inputtedQuery.length > maxQueryLength) {
        return getQueryTooLongErrorElement(
            className, inputtedQuery, maxQueryLength, errorState
        );
    } else {
        return null;
    }
}

function getErrorState(
    inputtedValue: string, maxValueLength: number, submittedErrorValue: boolean
): ErrorState {
    if (inputtedValue.length > 0 && inputtedValue.length <= maxValueLength) {
        return ErrorState.OK;
    }

    if (submittedErrorValue) {
        return ErrorState.Error;
    } else {
        return ErrorState.Warning;
    }
}

const SearchBarInput: React.FC<Props> = function(props) {
    const errorState = getErrorState(
        props.searchQuery, props.maxQueryLength, props.errorValueSubmitted
    );
    return (
        <React.Fragment>
            <div className='search-bar'>
                {useSearchInput(
                    props.searchQuery, props.onChange, props.loading,
                    errorState
                )}
                {useSearchClearButton(props.onChange, props.loading)}
                {getSearchSubmitButton(props.loading)}
            </div>
            {getErrorMessageElement(
                props.searchQuery, props.maxQueryLength, errorState
            )}
        </React.Fragment>
    );
};

export default SearchBarInput;
