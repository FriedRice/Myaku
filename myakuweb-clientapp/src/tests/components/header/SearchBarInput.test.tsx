/**
 * Tests for the SearchBarInput component.
 * @module tests/components/header/SearchBarInputInput.test
 */

import React from 'react';
import SearchBarInput from 'ts/components/header/SearchBarInput';

import {
    HTMLAttributes,
    ReactWrapper,
    mount,
} from 'enzyme';

function getInput(wrapper: ReactWrapper): ReactWrapper<HTMLAttributes> {
    const input = wrapper.find('input');
    expect(input).toHaveLength(1);
    return input;
}

function getClearButton(wrapper: ReactWrapper): ReactWrapper<HTMLAttributes> {
    const button = wrapper.find('button.search-clear');
    expect(button).toHaveLength(1);
    return button;
}

function getSubmitButton(wrapper: ReactWrapper): ReactWrapper<HTMLAttributes> {
    const button = wrapper.find('button.search-submit');
    expect(button).toHaveLength(1);
    return button;
}

function getErrorMessageText(
    wrapper: ReactWrapper, errorMessageClass: string
): string {
    const errorMessageElement = wrapper.find(`.${errorMessageClass}`);
    expect(errorMessageElement).toHaveLength(1);
    return errorMessageElement.text();
}


var wrapper = mount(<div />);
beforeEach(function() {
    wrapper = mount(
        <SearchBarInput
            searchQuery={'test'}
            loading={false}
            maxQueryLength={100}
            errorValueSubmitted={false}
            onChange={(): void => {}}
        />
    );
});

describe('<SearchBarInput />', function() {
    it('renders correctly', function() {
        expect(wrapper).toMatchSnapshot();
    });

    it('sets search query as input value', function() {
        expect(getInput(wrapper).props().value).toBe('test');

        wrapper.setProps({searchQuery: ''});
        expect(getInput(wrapper).props().value).toBe('');

        wrapper.setProps({searchQuery: 'backagain'});
        expect(getInput(wrapper).props().value).toBe('backagain');
    });

    it('disables input and buttons when loading', function() {
        expect(getInput(wrapper).props().readOnly).toBe(false);
        expect(getClearButton(wrapper).props().disabled).toBe(false);
        expect(getSubmitButton(wrapper).props().disabled).toBe(false);

        wrapper.setProps({loading: true});
        expect(getInput(wrapper).props().readOnly).toBe(true);
        expect(getClearButton(wrapper).props().disabled).toBe(true);
        expect(getSubmitButton(wrapper).props().disabled).toBe(true);

        wrapper.setProps({loading: false});
        expect(getInput(wrapper).props().readOnly).toBe(false);
        expect(getClearButton(wrapper).props().disabled).toBe(false);
        expect(getSubmitButton(wrapper).props().disabled).toBe(false);
    });

    it('shows loading spinner when loading', function() {
        expect(
            getSubmitButton(wrapper).find('.content-loading-query-spinner')
        ).toHaveLength(0);
        expect(getSubmitButton(wrapper).find('i.fa-search')).toHaveLength(1);

        wrapper.setProps({loading: true});
        expect(
            getSubmitButton(wrapper).find('.content-loading-query-spinner')
        ).toHaveLength(1);
        expect(getSubmitButton(wrapper).find('i.fa-search')).toHaveLength(0);

        wrapper.setProps({loading: false});
        expect(
            getSubmitButton(wrapper).find('.content-loading-query-spinner')
        ).toHaveLength(0);
        expect(getSubmitButton(wrapper).find('i.fa-search')).toHaveLength(1);
    });

    it('shows inputted query too long warning', function() {
        wrapper.setProps({maxQueryLength: 15});

        expect(wrapper.find('.input-warning-text')).toHaveLength(0);
        expect(getInput(wrapper).hasClass('warning-border')).toBe(false);

        wrapper.setProps({searchQuery: 'this-query-is-to-long'});
        const errorMessage = getErrorMessageText(
            wrapper, 'input-warning-text'
        );
        expect(errorMessage).toMatch(/Inputted/);
        expect(errorMessage).toMatch(/too long/);
        expect(errorMessage).toMatch(/21 \/ 15/);
        expect(getInput(wrapper).hasClass('warning-border')).toBe(true);

        wrapper.setProps({searchQuery: 'short-query'});
        expect(wrapper.find('.input-warning-text')).toHaveLength(0);
        expect(getInput(wrapper).hasClass('warning-border')).toBe(false);
    });

    it('shows submitted query too long error', function() {
        wrapper.setProps({maxQueryLength: 15});

        expect(wrapper.find('.input-error-text')).toHaveLength(0);
        expect(getInput(wrapper).hasClass('error-border')).toBe(false);

        wrapper.setProps({searchQuery: 'this-query-is-to-long'});
        expect(wrapper.find('.input-error-text')).toHaveLength(0);
        expect(getInput(wrapper).hasClass('error-border')).toBe(false);

        wrapper.setProps({errorValueSubmitted: true});
        const errorMessage = getErrorMessageText(wrapper, 'input-error-text');
        expect(errorMessage).toMatch(/Submitted/);
        expect(errorMessage).toMatch(/too long/);
        expect(errorMessage).toMatch(/21 \/ 15/);
        expect(getInput(wrapper).hasClass('error-border')).toBe(true);

        wrapper.setProps({errorValueSubmitted: false});
        expect(wrapper.find('.input-error-text')).toHaveLength(0);
        expect(getInput(wrapper).hasClass('error-border')).toBe(false);

        wrapper.setProps({searchQuery: 'short-query'});
        expect(wrapper.find('.input-error-text')).toHaveLength(0);
        expect(getInput(wrapper).hasClass('error-border')).toBe(false);
    });

    it('shows submitted empty query error', function() {
        wrapper.setProps({searchQuery: ''});

        expect(wrapper.find('.input-error-text')).toHaveLength(0);
        expect(getInput(wrapper).hasClass('error-border')).toBe(false);

        wrapper.setProps({errorValueSubmitted: true});
        const errorMessage = getErrorMessageText(wrapper, 'input-error-text');
        expect(errorMessage).toMatch(/No search query entered/);
        expect(getInput(wrapper).hasClass('error-border')).toBe(true);

        wrapper.setProps({errorValueSubmitted: false});
        expect(wrapper.find('.input-error-text')).toHaveLength(0);
        expect(getInput(wrapper).hasClass('error-border')).toBe(false);
    });

    it('ignores submitted error if query is valid', function() {
        wrapper.setProps({errorValueSubmitted: true});

        const input = getInput(wrapper);
        expect(input.hasClass('error-border')).toBe(false);
        expect(input.hasClass('warning-border')).toBe(false);

        expect(wrapper.find('.input-error-text')).toHaveLength(0);
        expect(wrapper.find('.input-warning-text')).toHaveLength(0);
    });

    it('calls onChange when input entered', function() {
        const mockOnChange = jest.fn();
        wrapper.setProps({onChange: mockOnChange});

        getInput(wrapper).getDOMNode<HTMLInputElement>().value = 'test2';
        getInput(wrapper).simulate('change');
        expect(mockOnChange).toBeCalledTimes(1);
        expect(mockOnChange).lastCalledWith('test2');

        getInput(wrapper).getDOMNode<HTMLInputElement>().value = 'test';
        getInput(wrapper).simulate('change');
        expect(mockOnChange).toBeCalledTimes(2);
        expect(mockOnChange).lastCalledWith('test');

        getInput(wrapper).getDOMNode<HTMLInputElement>().value = '';
        getInput(wrapper).simulate('change');
        expect(mockOnChange).toBeCalledTimes(3);
        expect(mockOnChange).lastCalledWith('');
    });

    it('calls onChange when input cleared', function() {
        const mockOnChange = jest.fn();
        wrapper.setProps({onChange: mockOnChange});

        getClearButton(wrapper).simulate('click');
        expect(mockOnChange).toBeCalledTimes(1);
        expect(mockOnChange).lastCalledWith('');

        getClearButton(wrapper).simulate('click');
        expect(mockOnChange).toBeCalledTimes(2);
        expect(mockOnChange).lastCalledWith('');
    });

    it('blurs focus on input clear button click', function() {
        const clearButtonElement = (
            getClearButton(wrapper).getDOMNode<HTMLButtonElement>()
        );
        clearButtonElement.focus();
        expect(document.activeElement).toBe(clearButtonElement);

        getClearButton(wrapper).simulate('click');
        expect(document.activeElement).not.toBe(clearButtonElement);
    });
});
