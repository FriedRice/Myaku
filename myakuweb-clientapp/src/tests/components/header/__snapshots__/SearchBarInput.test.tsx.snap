// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBarInput /> renders correctly 1`] = `
<SearchBarInput
  errorValueSubmitted={false}
  loading={false}
  maxQueryLength={100}
  onChange={[Function]}
  searchQuery="test"
>
  <div
    className="search-bar"
  >
    <input
      aria-label="Search input"
      className="search-input"
      id="search-input"
      name="q"
      onChange={[Function]}
      placeholder="Japanese word, set phrase, idiom, etc."
      readOnly={false}
      type="text"
      value="test"
    />
    <button
      aria-label="Search clear"
      className="search-clear"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa fa-times"
      />
    </button>
    <button
      aria-label="Search submit button"
      className="search-submit"
      disabled={false}
      type="submit"
    >
      <i
        className="fa fa-search"
      />
    </button>
  </div>
</SearchBarInput>
`;
